lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,259 @@ def render_encode_to_body(out, var, type, self_encode:)
         out.puts "  }"
       end
 
+      def render_struct_decode_from_interface(out, struct)
+        name = name(struct)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (s *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        declared_variables = []
+        struct.members.each do |m|
+          mn = name(m)
+          render_decode_from_body(out, "s.#{mn}", m.type, declared_variables: declared_variables, self_encode: false)
+        end
+        out.puts "  return n, nil"
+        out.puts "}"
+        out.break
+      end
+
+      def render_union_decode_from_interface(out, union)
+        name = name(union)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (u *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        render_decode_from_body(out, "u.#{name(union.discriminant)}", union.discriminant.type, declared_variables: [], self_encode: false)
+        switch_for(out, union, "u.#{name(union.discriminant)}") do |arm, kase|
+          out2 = StringIO.new
+          if arm.void?
+            out2.puts "// Void"
+          else
+            mn = name(arm)
+            type = arm.type
+            out2.puts "  u.#{mn} = new(#{reference arm.type})"
+            render_decode_from_body(out2, "(*u.#{mn})",type, declared_variables: [], self_encode: false)
+          end
+          out2.puts "  return n, nil"
+          out2.string
+        end
+        unless union.default_arm.present?
+          out.puts "  return n, fmt.Errorf(\"#{name(union.discriminant)} (#{reference union.discriminant.type}) switch value '%d' is not valid for union #{name}\", u.#{name(union.discriminant)})"
```

Comment:
❗ This error message looks a bit confusing once rendered. e.g.:
```go
fmt.Errorf("Type (EnvelopeType) switch value '%d' is not valid for union TransactionEnvelope", u.Type)
```
I suggest changing the order of the components so the error provides context first then zooms in:
```go
fmt.Errorf("union TransactionEnvelope has invalid Type (EnvelopeType) switch value '%d'", u.Type)
```
I think this can be achieved with:
```suggestion
          out.puts "  return n, fmt.Errorf(\"union #{name} has invalid #{name(union.discriminant)} (#{reference union.discriminant.type}) switch value '%d'\", u.#{name(union.discriminant)})"
```

Reply:
I was mimicking the error in go-xdr, but sure, I will change it